docs/howto/danceable-lyrics-input-tool/README.md

```diff
@@ -0,0 +1,495 @@
+
+Let's take a look at our final `on_complete`. It should look like something like this:
+
+```python
+def on_complete(self) -> None:
+    self.provider.io.info(f"{self.name} gathering sample lyrics...")
+
+    sample = (
+        pl.scan_csv(self.DATASETS_BASE + "genius_song_lyrics.csv")
+        .select(
+            pl.col("title").str.to_lowercase().alias("track_name"),
+            pl.col("artist").str.to_lowercase().alias("artist_name"),
+            "lyrics",
+            "language",
+            "views",
+        )
+        .filter(
+            pl.col("lyrics").str.contains("(?i)" + "|(?i)".join(self.LYRICS_TERMS))
+        )
+        .filter(pl.col("language") == "en")
+        .filter(pl.col("views") > self.MIN_VIEWS)
+        .collect()
+    )
+
+    self.provider.io.info(f"{self.name} gathering danceable track information...")
+
+    artists = (
+        pl.scan_csv(self.DATASETS_BASE + "artists.csv")
+        .select(
+            pl.col("name").str.to_lowercase().alias("artist_name"),
+            pl.col("id").alias("artist_id"),
+        )
+        .collect()
+    )
+
+    tracks = (
+        pl.scan_csv(self.DATASETS_BASE + "tracks.csv")
+        .select(
+            pl.col("name").str.to_lowercase().alias("track_name"),
+            pl.col("id").alias("track_id"),
+            "explicit",
+        )
+        .filter(pl.col("explicit") == 0)
+        .collect()
+    )
+
+    track_artists = (
+        pl.scan_csv(self.DATASETS_BASE + "r_track_artist.csv")
+        .select(pl.col("track_id"), pl.col("artist_id"))
+        .collect()
+    )
+
+    audio_features = (
+        pl.scan_csv(self.DATASETS_BASE + "audio_features.csv")
+        .select(pl.col("id").alias("track_id"), "danceability", "energy", "tempo",)
+        .filter(pl.col("danceability").is_between(*self.DANCEABILITY_RANGE))
+        .filter(pl.col("energy").is_between(*self.ENERGY_RANGE))
+        .filter(pl.col("tempo").is_between(*self.TEMPO_RANGE))
+        .collect()
+    )
+
+    danceability_tracks = (
+        track_artists.join(artists, on="artist_id")
+        .join(tracks, on="track_id")
+        .join(audio_features, on="track_id")
+    )
+
+    self.provider.io.info(f"{self.name} calculating final results...")
+
+    matches = pl.DataFrame()
+    for row in sample.rows(named=True):
+        artist = row["artist_name"]
+        track = row["track_name"]
+        m = (
+            danceability_tracks.select(
+                "artist_name",
+                "track_name",
+                "danceability",
+                "energy",
+                ("https://open.spotify.com/track/" + pl.col("track_id")),
+            )
+            .filter(pl.col("artist_name") == artist)
+            .filter(pl.col("track_name") == track)
+            .limit(1)
+        )
+
+        if not m.is_empty():
+            matches = pl.concat([matches, m])
+
+    matches = matches.sort("danceability", descending=True)
+
+    self.provider.write_to_anchor(
```

Comment:
Might want to explain what this is doing. I bet you folks in the Training will be confused since it might be the first time they are seeing Provider functions writing out.